File: DP(Dynamic_Programming, 동적계획법)/백준_1699_제곱수의 합/제곱수의 합.cs
using System.Text;
using System;
using static System.Console;

class Program
{
    static int[] arr = new int[100000];         //1부터 제곱수 들어갈 행렬
    public static void Main(string[] args)
    {
        int input = int.Parse(ReadLine());
        for(int i = 0; i<= input; i++)
        {
            arr[i] = i;
            for(int j = 0 ; j * j <= i; j++)
            {
                arr[i] = (arr[i] < (arr[i - j * j] + 1)) ? arr[i] : arr[i - j * j] + 1;
            }
        }
        Write(arr[input]);

    }
}

File: DP(Dynamic_Programming, 동적계획법)/백준_1699_제곱수의 합/제곱수의 합.py

x = int(input())

arr = [0 for _ in range(0, x+1)]

for i in range(0, x+1):
    arr[i] = i
    j = 0
    while j*j <= i:
        arr[i] = arr[i] if arr[i] < arr[i-j*j ]+1 else arr[i-j*j] +1
        j += 1

print(arr[x])

